packages/@aws-cdk/aws-servicecatalog/test/portfolio.test.ts

```diff
@@ -259,3 +259,167 @@ describe('Portfolio', () => {
     });
   });
 });
+
+describe('portfolio associations and product constraints', () => {
+  let app: cdk.App;
+  let stack: cdk.Stack;
+  let portfolio: servicecatalog.Portfolio;
+  let product: servicecatalog.CloudFormationProduct;
+  let testStackNameCounter = 0; //We need to maintain unique stack names since we use node addresses for ids
```

Comment:
Can you instead clear the maps in `AssociationManager` before each test?

Reply:
I am not sure if this is possible.  I had tried to find way to just pull the test name for each one and use it but that functionality is kind of limited.  Since the association manager is private anyway I feel like that would be not ideal.  I realize that an iterator is not fully ideal, but I couldn't find anything necessarily better or like global variables.

Reply:
Nothing is actually private in JavaScript 😃.

Take a look at [this PR](https://github.com/aws/aws-cdk/pull/14054) to see how you can achieve that (specifically, [here](https://github.com/aws/aws-cdk/pull/14054/files#diff-caff17cf05484c5f534825ee6c4f552c1bb3c3b38a7d7fd91ccebf1830a41701R2286)).

Reply:
Still would prefer to clear the maps instead.

Reply:
Working on this...

Reply:
I am not sure if this will work same way.  I don't think we should be adding code just for something to pass unit tests. We don't export the association manager, so I cannot access it.  Do you have anymore insight here?

Reply:
The PR I linked to shows you how to solve this problem:

```ts
import { ReplicaProvider } from '../lib/replica-provider';

// ...

  beforeEach(() => {
    (ReplicaProvider as any).getOrCreateCalls.clear();
  });
```

Reply:
It's worth thinking whether we need a static Map in the `AssociationManager` at all?

Constructs have an API that can check whether there exists a child with a given ID (`myConstruct.node.tryFindChild(id)`). Maybe that's good enough? You could get rid of the `scope` argument from the AssociationManager APIs, and simply always use either the `product`, or the `portfolio` as the parent of the created L1s.

Reply:
🤯

Reply:
see update, my guess is that the casting to `unknown` is still undesirable, but that was the fix recommended by IDE.

Reply:
Sorry, what update? I don't se anything.

Can we clear the maps here instead of the numbers ticking?

Reply:
argh I didn't realize I had 2 commits for set of comments when I requested review, I added the new commit that uses the nodes, it works perfectly well as far as I can tell...  cleans up the code (removes a lot ) and all the tests work plus I deployed actual things just to confirm.